File: examples/MIA_BPM/src/phadv.cc
#include <iostream>
#include <pcre.h>

#include "UAL/Common/Def.hh"
#include "UAL/APDF/APDF_Builder.hh"
#include "PAC/Beam/Position.hh"
#include "Optics/PacTMap.h"
#include "SMF/PacSmf.h"
#include "Main/Teapot.h"
#include "ZLIB/Tps/Space.hh"
#include "ual_sxf/Parser.hh"

int main(){

  ZLIB::Space space(6, 5);

   // ************************************************************************
  std::cout << "\n1. Read the SXF file" << std::endl;
  // ************************************************************************

  string sxfFile = "rhic_injection.sxf";

  string inFile   = "./../data/"; inFile   += sxfFile; 
  string echoFile = "./out/";     echoFile += sxfFile; echoFile += ".echo";
  string outFile =  "./out/";     outFile  += sxfFile; outFile  += ".out";

  UAL_SXF_Parser parser;
  parser.read(inFile.data(), echoFile.data()); 
  parser.write(outFile.data());

  // ************************************************************************
  std::cout << "\n2. Select a lattice for operations" << std::endl;
  // ************************************************************************

  string accName = "blue";

  PacLattices::iterator latIterator = PacLattices::instance()->find(accName);
  if(latIterator == PacLattices::instance()->end()){
    std::cerr << "There is no " + accName << " accelerator " << endl;
    exit(1);
  }

  PacLattice lattice = *latIterator;

  // ************************************************************************
  std::cout << "\n3. Define beam parameters." << std::endl;
  // ************************************************************************

  PAC::BeamAttributes ba;
  ba.setEnergy(250.0);

  // ************************************************************************
  std::cout << "\n4. Find closed orbit and Twiss parameters." << std::endl;
  // ************************************************************************

  Teapot teapot(lattice);

  // Find closed orbit

  PAC::Position clorbit;
  teapot.clorbit(clorbit, ba);

  std::cout << "closed orbit "
	    << " : x = " << clorbit.getX() 
	    << ", px = " << clorbit.getPX() 
	    << ", y = "  << clorbit.getY() 
	    << ", py = " << clorbit.getPY() 
	    << endl;

  // Find Twiss parameters in the injection point

  PacTwissData tw;
  teapot.twiss(tw, ba, clorbit);

  // ************************************************************************
  std::cout << "\n5. Make regular expression for selecting elements." << std::endl;
  // ************************************************************************

  std::string strPattern  = "Quadrupole"; // "^(q.*)";  

  const char *error;
  int erroffset;
  int overtor[30];   
  pcre* rePattern         = pcre_compile(strPattern.data(), 0, &error, &erroffset, NULL);
  pcre_extra* pePattern   = pcre_study(rePattern, 0, &error);

  // ************************************************************************
  std::cout << "\n6. Propagate Twiss parameters and print them for selected elements." << std::endl;
  // ************************************************************************

  ofstream outfile;
  outfile.open("quads.out");

  double at  = 0;
  tw.mu(0, 0.0);
  double mux = 0.0;

  tw.mu(1, 0.0);
  double muy = 0.0;

  int counter = 0;
  for(int i = 0; i < lattice.getNodeCount(); i++){

     UAL::AcceleratorNode* const anode = lattice.getNodeAt(i);  

     // Check the element design name

     std::string elname = anode->getDesignName();
     // int rc = pcre_exec(rePattern, pePattern, elname.data(), elname.size(), 
     // 0, 0, overtor, 30);
     std::string eltype = anode->getType();
     
     // if(rc > 1){
     // if(eltype == "Quadrupole") {
     if(eltype == "Vmonitor" || eltype == "Hmonitor" ) {
       outfile   << counter++ << " " 
		 << i << " " 
		 << elname << " " 
		 << at << " " 
		 << tw.mu(0)/UAL::pi/2.0 << " " 
		 << tw.beta(0) << "\n";
     }

     // find element map

     PacTMap sectorMap(6);
     sectorMap.refOrbit(clorbit);    
     teapot.trackMap(sectorMap, ba, i, i+1);

     // propagate Twiss parameters through this map

     teapot.trackTwiss(tw, sectorMap);

     // check mu

     if((tw.mu(0) - mux) < 0.0 ) { tw.mu(0, tw.mu(0) + 1.0); }
     mux = tw.mu(0);

     if((tw.mu(1) - muy) < 0.0 ) { tw.mu(1, tw.mu(1) + 1.0); }
     muy = tw.mu(1);

     // propagate closed orbit

     teapot.trackClorbit(clorbit, ba, i, i+1);    

     // update position

     at += anode->getLength();
  }

  outfile.close();
}

File: trunk/examples/MIA_BPM/phadv.pl
#!/usr/bin/perl

my $job_name   = "test_no_dxmp";

use File::Path;
use File::Basename;
mkpath(["./out/" . $job_name], 1, 0755);

# ------------------------------------------------------
# Create the ALE::UI::Shell  instance 
# ------------------------------------------------------

use lib ("$ENV{UAL_EXTRA}/ALE/api");
use ALE::UI::Shell;

print "Create the ALE::UI::Shell instance", "\n";

my $shell = new ALE::UI::Shell("print" => "./out/" . $job_name . "/log");

# ------------------------------------------------------
# Define the space of Taylor maps
# ------------------------------------------------------

print "Define the space of Taylor maps", "\n";

$shell->setMapAttributes("order" => 5);

# ------------------------------------------------------
# Define design elements and beam lines
# ------------------------------------------------------

# Read the SXF file

print "read the SXF file   ", time, "\n";

use lib ("$ENV{UAL_SXF}/api");
use UAL::SXF::Parser;
my $sxf_parser = new UAL::SXF::Parser();

$sxf_parser->read("./data/rhic_injection.sxf", "./out/" . $job_name . "/echo.sxf");

# ------------------------------------------------------
# Select and initialize a lattice for operations
# ------------------------------------------------------

print "Select and initialize a lattice", "\n";

# Select an accelerator for operations

$shell->use("lattice" => "blue");

# Write SMF into the FTPOT file

$shell->writeFTPOT("file" => "./out/" . $job_name . "/tpot");

# ------------------------------------------------------
# Define beam parameters
# ------------------------------------------------------

print "Define beam parameters", "\n";

$shell->setBeamAttributes("energy" => 250.0, "mass" => 0.93827231);

# ------------------------------------------------------
# Linear analysis
# ------------------------------------------------------

print "Linear analysis: ", "\n";

# Make general analysis
print " analysis\n";

$shell->analysis("print" => "./out/" . $job_name . "/analysis");

# Make linear matrix
print " matrix\n";

$shell->map("order" => 1, "print" => "./out/" . $job_name . "/map1"); 

# Calculate survey
print " survey\n";

$shell->survey("elements" => "", "print" => "./out/" . $job_name . "/survey"); 

# Calculate twiss
print " twiss\n";

# $shell->twiss("elements" => "cplmon", "print" => "./out/" . $job_name . "/twiss"); 

print_phadv($shell, "./phadv.out", "cplmon", 0.0);

print_phadv($shell, "./quads.out", "^(q.*)\$", 0.0);

sub print_phadv
{
    my ($shell, $file, $regex, $de) = @_;


    my $lattice = $shell->{lattice};
    my $code = $shell->{code};
    my $beam = $shell->{beam};

    my $dir = dirname($file);

    # Closed orbit

    my $orbit = new Pac::Position(); 
    $orbit->de($de);
    $code->clorbit($orbit, $beam);


    # One-Turn Map 

    my $oneTurnMap = new Pac::TMap(6);

    $oneTurnMap->refOrbit($orbit);	
    $code->map($oneTurnMap, $beam, 1); 


    my $eigenMap = new Pac::TMap(6);
    
    open(TWISS, ">$file") || die "can't create file $file";


    my $twiss = new Pac::TwissData; 
    my $chrom = new Pac::ChromData;
    my ($i, $le, $suml, $bName) = (0, 0, 0, " ");

    my $pi2grad = 360.0*1./atan2(1,1)/8.;   

 
    $code->chrom($chrom, $beam, $orbit);
    $twiss = $chrom->twiss();
    $twiss->mu(0, 0.0);
    $twiss->mu(1, 0.0);

    my $counter = 0;
    my $prev_phadv = $twiss->mu(0);
    my ($mux, $muy) = (0.0, 0.0);
 
    for($i=0; $i < $lattice->size; $i++){
	$le = $lattice->element($i);    
    
	if($le->genName =~ $regex) {
	  $output = sprintf("%5d %5d %-10s %10.3e %10.3e\n", 
			    $counter++, $i, $le->genName(), 
			    ($twiss->mu(0) - $prev_phadv)*$pi2grad,
			     $twiss->beta(0));		
	  print TWISS $output;
	  $prev_phadv = $twiss->mu(0);

	}

	# track map
        my $sectorMap = new Pac::TMap(6); 
        $sectorMap->refOrbit($orbit);  

	$code->trackMap($sectorMap, $beam, $i, $i+1);
	$code->trackClorbit($orbit, $beam, $i, $i+1);

	# track twiss
	$code->trackTwiss($twiss, $sectorMap);

	# check mu
	if(($twiss->mu(0) - $mux) < 0.0) { $twiss->mu(0, $twiss->mu(0) + 1.0); }
	$mux = $twiss->mu(0);
	if(($twiss->mu(1) - $muy) < 0.0) { $twiss->mu(1, $twiss->mu(1) + 1.0); }
	$muy = $twiss->mu(1);

	# update suml
    }   
   

    close(TWISS);
}


print "End", "\n";

1;
